### PR TITLE
refactor: port crypto function to Swift - WPB-10448

### PR DESCRIPTION
**Pull request by johnxnguyen**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-10448" title="WPB-10448" target="_blank">TaskWPB-10448</a>  [iOS] Renew access token if needed
  </td></table>
  <br />
 

### Issue

This PR ports some crypto methods from `NSData+ZMSCrypto.m` to Swift, placing them inside the new `WireFoundation` package. These crypto methods are needed to encrypt and decrypt authentication cookies at rest (inside the Keychain), which will come in another PR.

There are two types of encrypt/decrypt functions:
1. Encrypt some `Data` with a key, decrypt using the same key back into the original data.
2. Prefix some random data (IV) to the plaintext, encrypt, then decrypt and strip the prefix to return the original data.

Additionally, I've created a helper for generating random bytes, which is useful for generating new encryption keys.

### Testing

- Unit tests for AES encrypt/decrypt (both variants)
- Unit tests for random byte generation

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [ ] Make sure you use the API for UI elements that support large fonts.
- [ ] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [ ] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

File: WireFoundation/Sources/WireFoundation/Crypto/AES256Crypto.swift [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    /// - Parameters:
+    ///   - plaintext: The data to encrypt.
+    ///   - key: The encryption key. It must be 32 bytes.
+    ///
+    /// - Returns: The encrypted ciphertext.
+
+    public static func encryptAllAtOnceWithPrefixedIV(
+        plaintext: Data,
+        key: Data
+    ) throws -> Data {
```
Comment (netbe) [inline WireFoundation/Sources/WireFoundation/Crypto/AES256Crypto.swift]: would be a good idea to wrap this data into a custom type? as you don't know this data needs to be decrypted by `decryptAllAtOnceWithPrefixedIV`

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+//
+
+import CommonCrypto
+import XCTest
+
+@testable import WireFoundation
+
+final class AES256CryptoTests: XCTestCase {
+
+    // Plain ould encryption / decryption
```
Comment (netbe) [inline WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]: typo?

**Review by netbe — APPROVED** [review]
left some comments, but looks good!

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        let plaintext = try AES256Crypto.decryptAllAtOnceWithPrefixedIV(
+            ciphertext: ciphertext,
+            key: key
+        )
+
+        // Then
+        XCTAssertEqual(plaintext, originalData)
+    }
+
+    func testEncryptDecryptWithPrefixedIV_IVIsRandomlyGenerated() throws {
```
Comment (jullianm) [inline WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]:
```suggestion
    func testEncryptWithPrefixedIV_IVIsRandomlyGenerated() throws {
```

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        do {
+            // When
+            _ = try AES256Crypto.encryptAllAtOnce(
+                plaintext: originalData,
+                key: key
+            )
+
+            XCTFail("expected an error")
+
+        } catch AES256CryptoError.invalidKeyLength {
```
Comment (jullianm) [inline WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]: Maybe we can use `XCTAssertThrowsError` helper method from `WireTesting` in WireFoundation here

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        do {
+            // When
+            _ = try AES256Crypto.decryptAllAtOnce(
+                ciphertext: originalData,
+                key: key
+            )
+
+            XCTFail("expected an error")
+
+        } catch AES256CryptoError.invalidKeyLength {
```
Comment (jullianm) [inline WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]: here as well

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+@testable import WireFoundation
+
+final class AES256CryptoTests: XCTestCase {
+
+    // Plain ould encryption / decryption
+
+    // AES operates on fixed-size blocks of data, which for AES is 128 bits (16 bytes).
+    // Therefore we want to test encryption and decryption with data that is not a
+    // mutliple of the block size and with data that is. This ensures we test correct
```
Comment (jullianm) [inline WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]:
```suggestion
    // multiple of the block size and with data that is. This ensures we test correct
```

**Review by jullianm — APPROVED** [review]
looks good 👍

File: WireFoundation/Sources/WireFoundation/Crypto/AES256Crypto.swift (hunk 1 above)
Comment (johnxnguyen) [reply WireFoundation/Sources/WireFoundation/Crypto/AES256Crypto.swift]: done 6cf6efff5cd173ff242566c15629660d4ef1ffe8

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift (hunk 2 above)
Comment (johnxnguyen) [reply WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]: fixed d6661d2a94704f8619b7a20e03c6ec4c1f346cb8

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift (hunk 3 above)
Comment (johnxnguyen) [reply WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]: fixed 2279c1f6a7e505ca27bd47f0e8b7367957fe17e9

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift (hunk 4 above)
Comment (johnxnguyen) [reply WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]: done 6bf0dd395dac49f8b664ed435bacec1f65e2e22c

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift (hunk 5 above)
Comment (johnxnguyen) [reply WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]: done 6bf0dd395dac49f8b664ed435bacec1f65e2e22c

File: WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift (hunk 6 above)
Comment (johnxnguyen) [reply WireFoundation/Tests/WireFoundationTests/Crypto/AES256CryptoTests.swift]: fixed deff0adb46c4016163e342e298550a758d98231e

**Pull request merged**